app/jobs/download_manifest_job.rb

```
@@ -1,38 +1,62 @@
+  def get_service(_download)
+    fail "get_service must be implemented by a child class of DownloadManifestJob"
+  end
+
+  def service_name
+    fail "service_name must be implemented by a child class of DownloadManifestJob"
+  end
+
+  def client_error
+    fail "client_error must be implemented by a child class of DownloadManifestJob"
+  end
+
+  def manifest_fetched_at_name
+    "manifest_#{service_name}_fetched_at"
+  end
+
+  def connection_error_tag
+    "#{service_name}_connection_error"
+  end
+
+  # obtain docs from a service and save to the document model
+  # retuns <error>, <array of fetched documents>
   def perform(download)
-    external_documents = VBMSService.fetch_documents_for(download)
+    external_documents = download_from_service(download)
+    create_documents(download, external_documents) if !external_documents.empty?
+    return nil, external_documents
+  rescue client_error => e
+    capture_error(e)
+    return connection_error_tag, nil
+  end
 
-    if FeatureToggle.enabled?(:vva_service, user: download.user)
-      external_documents += VVAService.fetch_documents_for(download)
-    end
+  private
 
-    if external_documents.empty?
-      download.update_attributes!(status: :no_documents)
-      return
-    end
+  def capture_error(e)
+    Rails.logger.error "#{e.message}\n#{e.backtrace.join("\n")}"
+    Raven.capture_exception(e)
+  end
 
+  def create_documents(download, external_documents)
     download_documents = DownloadDocuments.new(
       download: download,
       external_documents: external_documents
     )
     download_documents.create_documents
-    download.update_attributes!(status: :pending_confirmation)
+  end
 
-  rescue VBMS::ClientError => e
-    capture_error(e, download, :vbms_connection_error)
-  rescue VVA::ClientError => e
-    capture_error(e, download, :vva_connection_error)
+  def download_from_service(download)
+    service = get_service(download)
+    return [] if !service
```

I think we should remove this line. First, we fail the get_service method if it's not implemented. Second, this could cause us to silently fail getting documents from one of the services. We should just allow it to error on the next line if this is nil.

---

Hmmm, okay I see you're returning nil for the VVA service if the feature flag isn't on. It works for now, but I think we should remove this when we turn it on for everyone.

---

Yes agreed @mdbenjam, it was purely because of the featureflag that I had to include this line.  Do we have a sense when we can turn VVA for everyone?